File: benchmarks/cpp/benchmark.cpp
#include <algorithm>
#include <iostream>
#include <unistd.h>
#include <fstream>
#include <string>
#include <chrono>
#include "trie.h"

std::vector<std::string> read_batch(std::ifstream &file, int batch_size) {
    std::vector<std::string> lines;

    for(int i = 0; i < batch_size; i++) {
        std::string line;
        
        if(!std::getline(file, line)) break;
        
        lines.push_back(line);
    }

    return lines;
}


Trie trie_insert_benchmark(std::string words_file, int batch_size = 1000000) {
    Trie trie;

    std::ifstream file(words_file, std::ifstream::in);
    int no_words = std::count(std::istreambuf_iterator<char>(file), std::istreambuf_iterator<char>(), '\n');
    int read_words = 0;

    file.clear();
    file.seekg(0);

    double total_duration = 0;

    while(true) {
        std::vector<std::string> batch = read_batch(file, batch_size);

        if(batch.empty()) break;

        auto start = std::chrono::high_resolution_clock::now();
        for(long unsigned int i = 0; i < batch.size(); i++) {
            trie.insert(batch.at(i));
        }
        auto stop = std::chrono::high_resolution_clock::now();

        double duration = (double) std::chrono::duration_cast<std::chrono::milliseconds>(stop - start).count();
        total_duration += duration;
        read_words += batch.size();
        
        std::cout << "\33[2K\r" << "Progress: " << 100*read_words / no_words << "%, " << batch.size() / duration << " words/ms";
        std::fflush(stdout);
    }
    
    file.close();

    std::cout << std::endl;
    std::cout << "Inserted " << read_words << " words in " << total_duration << "ms" << std::endl;
    std::cout << "Average inserts per millisecond (across batches): " << read_words / total_duration << std::endl;

    return trie;
}


void trie_search_benchmark(Trie trie, std::string words_file) {
    std::srand(std::time(nullptr));
    std::ifstream file(words_file, std::ifstream::in);
    int no_words = std::count(std::istreambuf_iterator<char>(file), std::istreambuf_iterator<char>(), '\n');
    
    file.clear();
    file.seekg(0);

    double total_duration = 0;
    int searched_words = 0;

    while(true) {
        std::vector<std::string> batch = read_batch(file, 1000000);
        if(batch.empty()) break;

        auto start = std::chrono::high_resolution_clock::now();
        for(long unsigned int i = 0; i < batch.size(); i++) {
            std::vector<std::string> res = trie.search(batch.at(i));
        }
        auto stop = std::chrono::high_resolution_clock::now();

        double duration = (double) std::chrono::duration_cast<std::chrono::milliseconds>(stop - start).count();
        total_duration += duration;
        searched_words += batch.size();
        
        std::cout << "\33[2K\r" << "Progress: " << 100*searched_words / no_words << "%, " << batch.size() / duration << " words/ms";
        std::fflush(stdout);
    }
    
    file.close();

    std::cout << std::endl;
    std::cout << "Searched for " << no_words << " words in " << total_duration << "ms" << std::endl;
    std::cout << "Average searches per millisecond (across batches): " << no_words / total_duration << std::endl;
}


double current_rss() {
    int tSize = 0, resident = 0, share = 0;
    std::ifstream buffer("/proc/self/statm");
    buffer >> tSize >> resident >> share;
    buffer.close();

    long page_size_kb = sysconf(_SC_PAGE_SIZE) / 1024; // in case x86-64 is configured to use 2MB pages
    double rss = resident * page_size_kb;
    return rss / 1024;
}


int main(int argc, char *argv[]) {
    if(argc <= 1) {
        std::cout << "Usage: " << argv[0] << " PATH_TO_WORDLIST" << std::endl;
        return -1;
    }
    
    double start_rss = current_rss();
    Trie trie = trie_insert_benchmark(std::string(argv[1]));
    std::cout << "Memory (RSS) used for Trie tree: " << current_rss() - start_rss << "mb" << std::endl;     
    trie_search_benchmark(trie, std::string(argv[1]));
    return 0;
}

File: benchmarks/python/benchmark.py
import argparse
import psutil
import time
import os


def read_chunks(words_file: str, words_file_encoding="utf-8", batch_size: int = 1000000):
    lines = []

    with open(words_file, "r", encoding=words_file_encoding) as f:
        for line in f:
            lines.append(line.strip())

            if len(lines) == batch_size:
                yield lines
                lines.clear()
    
    if len(lines) != 0:
        yield lines


def trie_insert_benchmark(words_file: str, implementation: str, batch_size: int = 1000000, words_file_encoding="utf-8", use_batch_insert=False):
    if implementation == "cpp":
        import trie
        tree = trie.Trie()
    else:
        from baseline.trie import Trie
        tree = Trie()

    duration = 0
    no_words = 0
    read_words = 0

    with open(words_file, "r", encoding=words_file_encoding) as f:
        for line in f:
            no_words += 1
    
    for batch in read_chunks(words_file, words_file_encoding=words_file_encoding, batch_size=batch_size):
        if use_batch_insert:
            t1 = time.time()
            tree.batch_insert(batch)
        else:
            t1 = time.time()
            for line in batch:
                tree.insert(line)

        t2 = time.time()
        
        batch_duration = (t2 - t1) * 1000
        duration += batch_duration
        read_words += len(batch)

        print(f"\33[2K\rProgress: {int(100*read_words/no_words)}%, {len(batch) / batch_duration} words/ms", end="")
    
    print("")
    print(f"Inserted {read_words} words in {duration}ms")
    print(f"Average inserts per millisecond (across batches): {read_words / duration}")

    return tree


def trie_search_benchmark(tree, words_file: str, words_file_encoding="utf-8"):
    duration = 0
    no_words = 0
    searched_words = 0

    with open(words_file, "r", encoding=words_file_encoding) as f:
        for line in f:
            no_words += 1
    
    for batch in read_chunks(words_file, words_file_encoding=words_file_encoding, batch_size=1000000):
        t1 = time.time()
        for line in batch:
            res = tree.search(line, False)
        t2 = time.time()
        
        batch_duration = (t2 - t1) * 1000
        duration += batch_duration
        searched_words += len(batch)

        print(f"\33[2K\rProgress: {int(100*searched_words/no_words)}%, {len(batch) / batch_duration} words/ms", end="")
    
    print("")
    print(f"Searched for {searched_words} words in {duration}ms")
    print(f"Average searches per millisecond (across batches): {searched_words / duration}")
                

if __name__ == "__main__":
    parser = argparse.ArgumentParser(prog="Trie Benchmark")
    parser.add_argument("--words", dest="words", help="Path to word list (single word per line)", required=True)
    parser.add_argument(
        "--implementation", dest="impl", choices=["cpp", "python"], 
        help="Implementation to use - either CPP via PyBind, or pure Python", default="python"
    )
    parser.add_argument("--words-encoding", dest="enc", help="Encoding to use when reading words file", default="utf-8")
    parser.add_argument("--use-batch-insert", dest="use_batch_insert", action="store_true", default=False)
    
    args = parser.parse_args()

    process = psutil.Process()
    start_mem = process.memory_info().rss
    tree = trie_insert_benchmark(args.words, args.impl, words_file_encoding=args.enc, use_batch_insert=args.use_batch_insert)
    cur_mem = process.memory_info().rss
    print(f"Memory (RSS) used for Trie tree: {(cur_mem - start_mem) / 1024 / 1024}mb")
    trie_search_benchmark(tree, args.words, words_file_encoding=args.enc)
